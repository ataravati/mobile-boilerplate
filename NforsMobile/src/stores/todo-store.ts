import { destroy, types, onSnapshot } from "mobx-state-tree";

export const Todo = types
  .model("Todo", {
    title: types.string,
    done: false,
  })
  .actions(self => ({
    toggle() {
      self.done = !self.done;
    },
  }));

const TodoStoreModel = types
  .model("TodoStore", {
    todos: types.array(Todo),
  })
  .actions(self => ({
    addTodo(text: string) {
      self.todos.push({
        title: text,
        done: false,
      } as typeof Todo.Type);
    },
    destroyTodo(todo: typeof Todo.Type) {
      destroy(todo);
    },
  }));

export const todoStore = TodoStoreModel.create({
  todos: [],
});

onSnapshot(todoStore, snapshot => {
  console.log(snapshot);
});

export type TodoStore = typeof TodoStoreModel.Type;
