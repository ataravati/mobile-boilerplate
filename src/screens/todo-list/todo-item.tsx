import React from "react";
import { StyleSheet } from "react-native";

import { Button, Text, View } from "native-base";

export function TodoItem({ destroyTodo, key, todo }) {
  return (
    <View key={key} style={styles.todo} testID="todo-item">
      <Text testID="todo-title">{todo.title}</Text>
      <Text testID="todo-completed">{todo.completed}</Text>
      <View>
        <Button onPress={() => destroyTodo(todo)}>
          <Text>Delete</Text>
        </Button>
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  todo: {
    flex: 1,
    flexDirection: "row",
    justifyContent: "space-between",
    margin: 10,
  },
});
