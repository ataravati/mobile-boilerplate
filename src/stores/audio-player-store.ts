import { types, onSnapshot, flow } from "mobx-state-tree";
import Sound from "react-native-sound";
import { Platform } from "react-native";

Sound.setCategory("Playback", true);
if (Platform.OS === "ios") {
  Sound.setMode("SpokenAudio");
}

const sounds = {};

function loadSound(filename: string) {
  return new Promise(function(resolve, reject) {
    const sound = new Sound(filename, "", function(error) {
      if (error) {
        reject(error);
      } else {
        resolve(sound);
      }
    });
  });
}

function play(filename: string) {
  return new Promise(function(resolve, reject) {
    sounds[filename].play(function(success?: boolean) {
      if (success === true) {
        resolve();
      } else {
        reject();
      }
    });
  });
}

function getCurrentTime(filename: string) {
  return new Promise(function(resolve, reject) {
    sounds[filename].getCurrentTime(function(
      seconds: number,
      isPlaying: boolean,
    ) {
      if (isPlaying === true) {
        resolve(seconds);
      } else {
        reject();
      }
    });
  });
}

const AudioPlayerStoreModel = types
  .model("AudioPlayerStore", {
    isLoading: types.boolean,
    paused: types.boolean,
    filename: types.string,
    duration: types.number,
    currentTime: types.number,
  })
  .actions(self => ({
    load: flow(function*(filename: string) {
      self.isLoading = true;
      try {
        sounds[filename] = yield loadSound(filename);
        self.isLoading = false;
        self.filename = filename;
        self.duration = sounds[filename].getDuration();
      } catch (error) {
        console.log("Failed to load the audio file.", error);
      }
    }),
    setPaused(paused: boolean) {
      self.paused = paused;
    },
    play: flow(function*() {
      if (sounds[self.filename].isLoaded()) {
        self.paused = false;
        if (Platform.OS === "ios") Sound.setActive(true);

        try {
          yield play(self.filename);
          self.paused = true;
          if (Platform.OS === "ios") Sound.setActive(false);
        } catch {
          if (Platform.OS === "android") sounds[self.filename].reset();
        }
      }
    }),
    pause() {
      sounds[self.filename].pause();
      self.paused = true;
    },
    updateCurrentTime: flow(function*() {
      try {
        let seconds = yield getCurrentTime(self.filename);
        seconds = seconds > self.duration ? self.duration : seconds;
        self.currentTime = seconds;
      } catch {}
    }),
    setCurrentTime(time: number) {
      sounds[self.filename].setCurrentTime(time);
      self.currentTime = time;
    },
  }));

export const audioPlayerStore = AudioPlayerStoreModel.create({
  isLoading: true,
  paused: true,
  filename: "",
  duration: -1,
  currentTime: 0,
});

onSnapshot(audioPlayerStore, snapshot => {
  console.log("snapshot:", snapshot);
});

export type AudioPlayerStore = typeof AudioPlayerStoreModel.Type;
