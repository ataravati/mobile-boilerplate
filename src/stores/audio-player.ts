import Sound from "react-native-sound";
import { Platform } from "react-native";

Sound.setCategory("Playback", true);
if (Platform.OS === "ios") {
  Sound.setMode("SpokenAudio");
}

const sounds = {};

export default class AudioPlayer {
  path = "";

  load = (path: string, time: number = 0, volume: number = 1) => {
    const that = this;
    return new Promise((resolve, reject) => {
      const basePath =
        path.startsWith("http") === true ? "" : Sound.MAIN_BUNDLE;
      const sound = new Sound(path, basePath, error => {
        if (error) {
          reject(error);
        } else {
          if (time > 0) sound.setCurrentTime(time);
          sound.setVolume(volume);
          sounds[path] = sound;
          that.path = path;
          resolve();
        }
      });
    });
  };

  isLoaded = () => {
    return this.path ? sounds[this.path].isLoaded() : false;
  };

  play = () => {
    return new Promise((resolve, reject) => {
      if (Platform.OS === "ios") Sound.setActive(true);
      sounds[this.path].play((success?: boolean) => {
        if (success === true) {
          if (Platform.OS === "ios") Sound.setActive(false);
          resolve();
        } else {
          if (Platform.OS === "android") sounds[this.path].reset();
          reject();
        }
      });
    });
  };

  pause = () => {
    return new Promise(resolve => {
      sounds[this.path].pause(() => {
        resolve();
      });
    });
  };

  getDuration = () => {
    return sounds[this.path].getDuration();
  };

  getCurrentTime = () => {
    return new Promise(resolve => {
      sounds[this.path].getCurrentTime(
        (seconds: number, isPlaying: boolean) => {
          resolve({ seconds, isPlaying });
        },
      );
    });
  };

  setCurrentTime = (time: number) => {
    sounds[this.path].setCurrentTime(time);
  };

  release = () => {
    sounds[this.path].release();
    sounds[this.path] = null;
  };

  getVolume = () => {
    return sounds[this.path].getVolume();
  };

  setVolume = (volume: number) => {
    sounds[this.path].setVolume(volume);
  };
}
