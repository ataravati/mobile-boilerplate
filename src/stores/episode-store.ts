import { Platform } from "react-native";
import { flow, types, onSnapshot } from "mobx-state-tree";
import RNFetchBlob from "rn-fetch-blob";
import { getPodcasts, updateEpisode } from "../database/allSchemas";

export const Episode = types
  .model("Episode", {
    id: types.number,
    title: types.string,
    url: types.string,
    localPath: types.maybe(types.string),
    isLocal: types.optional(types.boolean, false),
    isDownloading: types.optional(types.boolean, false),
  })
  .views(self => ({
    get path() {
      return self.isLocal === true && self.localPath
        ? self.localPath
        : self.url;
    },
  }))
  .actions(self => ({
    download: flow(function*() {
      self.isDownloading = true;
      const dirs = RNFetchBlob.fs.dirs;
      const filename = self.url.split("/").pop() as string;
      try {
        const res = yield RNFetchBlob.config({
          fileCache: true,
          path: dirs.MainBundleDir + "/" + filename,
        }).fetch("GET", self.url);
        console.log(`The file saved to ${res.path()}.`);
        self.isLocal = true;
        self.localPath = Platform.OS === "android" ? res.path() : filename;

        yield updateEpisode(self);
      } catch (error) {
        console.log(`Could not download ${self.url}.`, error);
      }
      self.isDownloading = false;
    }),
  }));

const EpisodeStoreModel = types
  .model("EpisodeStore", {
    isLoading: types.boolean,
    episodes: types.array(Episode),
  })
  .actions(self => ({
    fetchAll: flow(function*() {
      self.isLoading = true;
      try {
        const podcasts = yield getPodcasts();
        const episodes = podcasts[0].episodes.map(episode => {
          return {
            id: episode.id,
            title: episode.title,
            url: episode.url,
            localPath: episode.localPath,
            isLocal: episode.isLocal,
          };
        });
        self.episodes = episodes;
      } catch {}
      self.isLoading = false;
    }),
    downloadEpisode: flow(function*(episode: typeof Episode.Type) {
      episode.download();
    }),
  }));

export const episodeStore = EpisodeStoreModel.create({
  isLoading: false,
  episodes: [],
});

onSnapshot(episodeStore, snapshot => {
  console.log("snapshot:", snapshot);
});

export type EpisodeStore = typeof EpisodeStoreModel.Type;
