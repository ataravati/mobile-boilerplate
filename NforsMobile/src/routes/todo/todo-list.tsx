import React from "react";
import { inject, observer } from "mobx-react/native";
import {
  Button,
  Container,
  Content,
  Footer,
  Form,
  Header,
  Icon,
  Input,
  Item,
  Text,
  View,
} from "native-base";
import { StyleSheet } from "react-native";

import { TodoStore, Todo } from "../../stores/todo-store";

interface TodoListProps {
  todoStore?: TodoStore;
}

interface TodoListState {
  newTodoText: string;
}

export const TodoItem = ({ destroyTodo, key, todo }) => (
  <View key={key} style={styles.todo}>
    <Text>{todo.title}</Text>
    <Text>{todo.done}</Text>
    <View>
      <Button onPress={() => destroyTodo(todo)}>
        <Text>Delete</Text>
      </Button>
    </View>
  </View>
);

const TodoItems = ({
  destroyTodo,
  todos,
}: {
  destroyTodo: (todo: typeof Todo.Type) => void;
  todos: typeof Todo.Type[];
}) => {
  return (
    <View style={styles.todos}>
      {todos.map((todo: typeof Todo.Type, index: number) => (
        <TodoItem destroyTodo={destroyTodo} key={index} todo={todo} />
      ))}
    </View>
  );
};

@inject("todoStore")
@observer
export class TodoList extends React.Component<TodoListProps, TodoListState> {
  state = {
    newTodoText: "",
  };

  destroyTodo = (todo: typeof Todo.Type) => {
    this.props.todoStore!.destroyTodo(todo);
  };

  addTodo = (text: string) => {
    this.props.todoStore!.addTodo(text);
  };

  render() {
    return (
      <Container>
        <Header />
        <Content>
          <TodoItems
            todos={this.props.todoStore!.todos.slice()}
            destroyTodo={this.destroyTodo}
          />
          <View style={styles.newTodo}>
            <View>
              <Form>
                <Item>
                  <Icon active name="create" />
                  <Input
                    placeholder="New Todo..."
                    onChangeText={text => this.setState({ newTodoText: text })}
                    value={this.state.newTodoText || ""}
                  />
                </Item>
              </Form>
              <Button
                full
                onPress={() => {
                  this.addTodo(this.state.newTodoText);
                  this.setState({
                    newTodoText: "",
                  });
                }}
              >
                <Text>Add Todo</Text>
              </Button>
            </View>
          </View>
        </Content>
        <Footer />
      </Container>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: "column",
    justifyContent: "space-between",
  },
  content: {
    margin: 10,
    flex: 1,
    flexDirection: "column",
  },
  newTodo: {
    flex: 1,
  },
  todos: {
    flex: 3,
    flexDirection: "column",
  },
  todo: {
    flex: 1,
    flexDirection: "row",
    justifyContent: "space-between",
    margin: 10,
  },
});
