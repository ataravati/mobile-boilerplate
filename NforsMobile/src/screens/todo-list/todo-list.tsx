import React from "react";
import { inject, observer } from "mobx-react/native";
import {
  Button,
  Container,
  Content,
  Footer,
  Form,
  Header,
  Icon,
  Input,
  Item,
  Text,
  View,
} from "native-base";
import { StyleSheet } from "react-native";
import { TodoItem } from "./todo-item";
import { createStackNavigator } from "react-navigation";

import { TodoStore, Todo } from "../../stores/todo-store";

interface TodoListProps {
  todoStore?: TodoStore;
}

interface TodoListState {
  newTodoText: string;
}

const TodoItems = ({
  destroyTodo,
  todos,
}: {
  destroyTodo: (todo: typeof Todo.Type) => void;
  todos: typeof Todo.Type[];
}) => {
  return (
    <View style={styles.todos}>
      {todos.map((todo: typeof Todo.Type, index: number) => (
        <TodoItem destroyTodo={destroyTodo} key={index} todo={todo} />
      ))}
    </View>
  );
};

@inject("todoStore")
@observer
class TodoListComponent extends React.Component<TodoListProps, TodoListState> {
  state = {
    newTodoText: "",
  };

  destroyTodo = (todo: typeof Todo.Type) => {
    this.props.todoStore!.destroyTodo(todo);
  };

  addTodo = (text: string) => {
    this.props.todoStore!.addTodo(text);
  };

  render() {
    return (
      <Container>
        <Header />
        <Content>
          <TodoItems
            todos={this.props.todoStore!.todos.slice()}
            destroyTodo={this.destroyTodo}
          />
          <View style={styles.newTodo}>
            <View>
              <Form>
                <Item>
                  <Icon active name="create" />
                  <Input
                    placeholder="New Todo..."
                    onChangeText={text => this.setState({ newTodoText: text })}
                    value={this.state.newTodoText || ""}
                  />
                </Item>
              </Form>
              <Button
                full
                onPress={() => {
                  this.addTodo(this.state.newTodoText);
                  this.setState({
                    newTodoText: "",
                  });
                }}
              >
                <Text>Add Todo</Text>
              </Button>
              <Button
                full
                onPress={() =>
                  this.props.navigation.navigate("PointlessScreen")
                }
              >
                <Text>Go to a pointless screen</Text>
              </Button>
            </View>
          </View>
        </Content>
        <Footer />
      </Container>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: "column",
    justifyContent: "space-between",
  },
  content: {
    margin: 10,
    flex: 1,
    flexDirection: "column",
  },
  newTodo: {
    flex: 1,
  },
  todos: {
    flex: 3,
    flexDirection: "column",
  },
});

export const TodoList = createStackNavigator({
  TodoList: {
    screen: TodoListComponent,
  },
});
