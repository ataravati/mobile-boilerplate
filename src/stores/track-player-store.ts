import { types, onSnapshot, flow, getSnapshot } from "mobx-state-tree";
import TrackPlayer, { Track } from "react-native-track-player";
import { Episode } from "./episode-store";

const snapshots = {};

const onStateChanged = TrackPlayer.addEventListener("playback-state", data => {
  console.log("State:", data.state);
  audioPlayerStore.changeState(data.state);
});

const AudioPlayerStoreModel = types
  .model("AudioPlayerStore", {
    state: types.maybe(types.string),
    duration: types.number,
    currentTime: types.number,
    volume: types.number,
    speed: types.number,
    episode: types.maybe(types.reference(Episode)),
  })
  .views(self => ({
    get isLoading() {
      return self.state === TrackPlayer.STATE_BUFFERING;
    },
    get isPlaying() {
      return self.state === TrackPlayer.STATE_PLAYING;
    },
    get paused() {
      return self.state === TrackPlayer.STATE_PAUSED;
    },
  }))
  .actions(self => ({
    setup: flow(function*() {
      console.log("STATE_NONE", TrackPlayer.STATE_NONE);
      console.log("STATE_BUFFERING", TrackPlayer.STATE_BUFFERING);
      console.log("STATE_PLAYING", TrackPlayer.STATE_PLAYING);
      console.log("STATE_PAUSED", TrackPlayer.STATE_PAUSED);
      console.log("STATE_STOPPED", TrackPlayer.STATE_STOPPED);

      yield TrackPlayer.setupPlayer();
      TrackPlayer.updateOptions({
        stopWithApp: true,
        capabilities: [
          // TrackPlayer.CAPABILITY_PLAY,
          // TrackPlayer.CAPABILITY_PAUSE,
        ],
        compactCapabilities: [
          // TrackPlayer.CAPABILITY_PLAY,
          // TrackPlayer.CAPABILITY_PAUSE,
        ],
        notificationCapabilities: [],
      });
    }),
    addEpisode: flow(function*(episode: typeof Episode.Type) {
      // If the track has changed...
      if (self.episode && episode.path !== self.episode.path) {
        self.currentTime = yield TrackPlayer.getPosition();
        snapshots[self.episode.path] = getSnapshot(self);
        yield TrackPlayer.reset();

        self.speed = 1;
        if (snapshots[episode.path]) {
          self.currentTime = snapshots[episode.path].currentTime;
          self.volume = snapshots[episode.path].volume;
        } else {
          self.currentTime = 0;
          self.volume = 1;
        }
      }

      // If it's a new track...
      if (!self.episode || self.episode.path !== episode.path) {
        const track = {
          id: episode.id.toString(),
          url: episode.path,
          title: episode.title,
          artist: "رادیو روغن حبه‌ی انگور",
        };

        yield TrackPlayer.add(track);

        self.episode = episode;
      }
    }),
    changeState: flow(function*(state: string) {
      self.state = state;

      if (self.state === TrackPlayer.STATE_PLAYING) {
        self.duration = yield TrackPlayer.getDuration();
        yield TrackPlayer.seekTo(self.currentTime);
        yield TrackPlayer.setVolume(self.volume);
      }

      if (self.episode) {
        (self.episode as typeof Episode.Type).changeLoadingStatus(
          self.isLoading,
        );

        (self.episode as typeof Episode.Type).changePlayingStatus(
          self.isPlaying,
        );
      }
    }),
    play: flow(function*() {
      try {
        yield TrackPlayer.play();
      } catch {}
    }),
    pause: flow(function*() {
      try {
        yield TrackPlayer.pause();
      } catch {}
    }),
    updateCurrentTime: flow(function*() {
      try {
        let currentTime = yield TrackPlayer.getPosition();
        currentTime = currentTime > self.duration ? self.duration : currentTime;
        self.currentTime = currentTime;
      } catch {}
    }),
    setCurrentTime: flow(function*(time: number) {
      yield TrackPlayer.seekTo(time);
      self.currentTime = time;
    }),
    setVolume: flow(function*(volume: number) {
      yield TrackPlayer.setVolume(volume);
      self.volume = volume;
    }),
    setSpeed: flow(function*(value: number) {
      yield TrackPlayer.setRate(value);
      self.speed = value;
    }),
  }));

export const audioPlayerStore = AudioPlayerStoreModel.create({
  state: TrackPlayer.STATE_NONE,
  duration: -1,
  currentTime: 0,
  volume: 1,
  speed: 1,
});

onSnapshot(audioPlayerStore, snapshot => {
  // console.log("snapshot:", snapshot);
});

export type AudioPlayerStore = typeof AudioPlayerStoreModel.Type;
