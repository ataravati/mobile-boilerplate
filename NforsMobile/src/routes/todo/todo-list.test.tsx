import React from "react";
import renderer from "react-test-renderer";
import { Provider } from "mobx-react";

import { TodoList } from "./todo-list";
import { todoStore } from "../../stores/todo-store";

describe("todo-list", () => {
  it("can render snapshot", () => {
    const tree = renderer
      .create(
        <Provider todoStore={todoStore}>
          <TodoList />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("can render", () => {
    const tree = renderer
      .create(
        <Provider todoStore={todoStore}>
          <TodoList />
        </Provider>,
      )
      .toJSON();
  });
});
