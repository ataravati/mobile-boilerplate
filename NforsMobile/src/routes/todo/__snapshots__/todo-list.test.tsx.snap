// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todo-list can render snapshot 1`] = `
<react-native-mock>
  <react-native-mock>
    <react-native-mock />
    <react-native-mock />
  </react-native-mock>
  <react-native-mock>
    <react-native-mock />
    <react-native-mock>
      <react-native-mock>
        <react-native-mock>
          <react-native-mock>
            <react-native-mock>
              ï…¥
            </react-native-mock>
            <react-native-mock />
          </react-native-mock>
        </react-native-mock>
        <react-native-mock>
          <react-native-mock>
            Add Todo
          </react-native-mock>
        </react-native-mock>
      </react-native-mock>
    </react-native-mock>
  </react-native-mock>
  <react-native-mock />
</react-native-mock>
`;
